gnovm/cmd/gno/repl.go

```diff
@@ -91,68 +113,196 @@ func runRepl(cfg *replCfg) error {
 		handleInput(r, cfg.initialCommand)
 	}
 
-	fmt.Fprint(os.Stdout, "gno> ")
+	fmt.Fprint(os.Stdout, gnoREPL)
+
+	var (
+		inEdit      bool
+		prev        string
+		indentLevel int
+	)
 
-	inEdit := false
-	prev := ""
 	liner := bufio.NewScanner(os.Stdin)
 
 	for liner.Scan() {
 		line := liner.Text()
 
-		if l := strings.TrimSpace(line); l == ";" {
-			line, inEdit = "", false
-		} else if l == "/editor" {
-			line, inEdit = "", true
-			fmt.Fprintln(os.Stdout, "// enter a single ';' to quit and commit")
-		}
+		trimmedLine := strings.TrimSpace(line)
+
+		indentLevel = updateIndentLevel(trimmedLine, indentLevel)
+		line, inEdit = handleEditor(line)
+
 		if prev != "" {
 			line = prev + "\n" + line
 			prev = ""
 		}
+
 		if inEdit {
-			fmt.Fprint(os.Stdout, "...  ")
+			fmt.Fprint(os.Stdout, inEditMode)
 			prev = line
+
 			continue
 		}
 
 		if err := handleInput(r, line); err != nil {
 			var goScanError scanner.ErrorList
 			if errors.As(err, &goScanError) {
-				// We assune that a Go scanner error indicates an incomplete Go statement.
+				// We assume that a Go scanner error indicates an incomplete Go statement.
 				// Append next line and retry.
 				prev = line
 			} else {
 				fmt.Fprintln(os.Stderr, err)
 			}
 		}
 
-		if prev == "" {
-			fmt.Fprint(os.Stdout, "gno> ")
-		} else {
-			fmt.Fprint(os.Stdout, "...  ")
-		}
+		printPrompt(indentLevel, prev)
 	}
+
 	return nil
 }
 
+func handleEditor(line string) (string, bool) {
+	if l := strings.TrimSpace(line); l == ";" {
+		return "", false
+	} else if l == editorCommand {
+		fmt.Fprintln(os.Stdout, "// enter a single ';' to quit and commit")
+		return "", true
+	}
+
+	return line, false
+}
+
+func updateIndentLevel(line string, indentLevel int) int {
+	openCount, closeCount := countBrackets(line)
+	increaseIndent := shouldIncreaseIndent(line)
+	indentLevel += openCount - closeCount
+
+	if indentLevel < 0 {
+		indentLevel = 0
+	}
+
+	if increaseIndent {
+		indentLevel++
+	}
+
+	return indentLevel
+}
+
+func countBrackets(line string) (int, int) {
+	openCount, closeCount := 0, 0
+	inString, inComment, inSingleLineComment := false, false, false
+	var stringChar rune
+
+	for i, char := range line {
+		if !inString && !inComment && !inSingleLineComment {
+			switch char {
+			case '{', '(', '[':
+				openCount++
+			case '}', ')', ']':
+				closeCount++
+			case '"', '\'':
+				inString = true
+				stringChar = char
+			case '/':
+				if i < len(line)-1 {
+					if line[i+1] == '/' {
+						inSingleLineComment = true
+					} else if line[i+1] == '*' {
+						inComment = true
+					}
+				}
+			}
+		} else if inString && char == stringChar {
+			inString = false
+		} else if inComment && i < len(line)-1 && char == '*' && line[i+1] == '/' {
+			inComment = false
+		}
+	}
+
+	return openCount, closeCount
+}
+
+func shouldIncreaseIndent(line string) bool {
```

Comment:
I think this should be removed unless there is someway this can return true outside of testing cases.  It seems like this is only ever called from from one spot and when the line is passed to that one spot, it has come from the scanner which already trims off any trailing newline character. The line also has all white space trimmed, so I'm not sure in which case this could return true when it is used as intended from the command line.

Reply:
I check the `shouldIncreaseIndent` function actually only returns `false` i practice. Therefore, I completely removed this functions